components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -0,0 +1,184 @@
+#include "base/metrics/histogram_macros.h"
+#include "base/no_destructor.h"
+#include "base/task/post_task.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/body_sniffer/body_sniffer_url_loader.h"
+#include "brave/components/de_amp/browser/de_amp_service.h"
+#include "brave/components/de_amp/browser/de_amp_throttle.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/mojom/early_hints.mojom.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+
+namespace de_amp {
+
+namespace {
+
+constexpr uint32_t kReadBufferSize = 65536;
+
+}  // namespace
+
+// static
+std::tuple<mojo::PendingRemote<network::mojom::URLLoader>,
+           mojo::PendingReceiver<network::mojom::URLLoaderClient>,
+           DeAmpURLLoader*>
+DeAmpURLLoader::CreateLoader(
+    base::WeakPtr<body_sniffer::BodySnifferThrottle> throttle,
+    const GURL& response_url,
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    DeAmpService* service,
+    content::WebContents* contents) {
+  mojo::PendingRemote<network::mojom::URLLoader> url_loader;
+  mojo::PendingRemote<network::mojom::URLLoaderClient> url_loader_client;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient>
+      url_loader_client_receiver =
+          url_loader_client.InitWithNewPipeAndPassReceiver();
+
+  auto loader = base::WrapUnique(new DeAmpURLLoader(
+      std::move(throttle), response_url, std::move(url_loader_client),
+      std::move(task_runner), service, contents));
+  DeAmpURLLoader* loader_rawptr = loader.get();
+  mojo::MakeSelfOwnedReceiver(std::move(loader),
+                              url_loader.InitWithNewPipeAndPassReceiver());
+  return std::make_tuple(std::move(url_loader),
+                         std::move(url_loader_client_receiver), loader_rawptr);
+}
+
+DeAmpURLLoader::DeAmpURLLoader(
+    base::WeakPtr<body_sniffer::BodySnifferThrottle> throttle,
+    const GURL& response_url,
+    mojo::PendingRemote<network::mojom::URLLoaderClient>
+        destination_url_loader_client,
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    DeAmpService* service,
+    content::WebContents* contents)
+    : body_sniffer::BodySnifferURLLoader(
+          throttle,
+          response_url,
+          std::move(destination_url_loader_client),
+          task_runner),
+      contents_(contents),
+      de_amp_service(service) {}
+
+DeAmpURLLoader::~DeAmpURLLoader() = default;
+
+void DeAmpURLLoader::OnBodyReadable(MojoResult) {
+  if (state_ == State::kSending) {
+    // The pipe becoming readable when kSending means all buffered body has
+    // already been sent.
+    ForwardBodyToClient();
+    return;
+  }
+  if (!BodySnifferURLLoader::CheckBufferedBody(kReadBufferSize)) {
+    return;
+  }
+
+  DeAmp();
+
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void DeAmpURLLoader::DeAmp() {
```

Comment:
nit: I would probably call this method something like `MaybeRedirectToCanonicalLink`